File: Mooga/src/de/lumpn/mooga/ranking/CrowdingDistanceRanking.java
package de.lumpn.mooga.ranking;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import de.lumpn.mooga.Individual;
import de.lumpn.mooga.ranking.impl.CrowdingDistanceComparator;
import de.lumpn.mooga.ranking.impl.CrowdingDistanceIndividual;
import de.lumpn.mooga.ranking.impl.DominationComparator;
import de.lumpn.mooga.ranking.impl.ScoreComparator;

public final class CrowdingDistanceRanking implements Ranking {

	@Override
	public List<Individual> rank(List<Individual> individuals) {

		// trivial?
		if (individuals.size() < 2) {
			return new ArrayList<Individual>(individuals);
		}

		// split into non-dominated and dominated
		List<Individual> nonDominated = new ArrayList<Individual>();
		List<Individual> dominated = new ArrayList<Individual>();
		for (Individual individual : individuals) {

			// check domination
			boolean isDominated = false;
			for (Individual other : individuals) {
				if (other == individual) continue;
				if (comparator.compare(individual, other) < 0) {
					isDominated = true;
					break;
				}
			}

			// put in respective list
			if (isDominated) {
				dominated.add(individual);
			} else {
				nonDominated.add(individual);
			}
		}

		// sort non-dominated by crowding distance
		List<Individual> sortedNonDominated = sortByCrowdingDistance(nonDominated);

		// recursively rank the dominated individuals
		List<Individual> rankedDominated = rank(dominated);

		// concatenate both lists
		List<Individual> result = new ArrayList<Individual>();
		result.addAll(sortedNonDominated);
		result.addAll(rankedDominated);
		return result;
	}

	private List<Individual> sortByCrowdingDistance(List<Individual> individuals) {

		// trivial?
		if (individuals.size() < 2) {
			return new ArrayList<Individual>(individuals);
		}

		// create crowding distance wrapper
		List<CrowdingDistanceIndividual> wrappedIndividuals = new ArrayList<CrowdingDistanceIndividual>();
		for (Individual individual : individuals) {
			wrappedIndividuals.add(new CrowdingDistanceIndividual(individual));
		}

		// calculate each crowding distance
		final int lastIndex = wrappedIndividuals.size() - 1;
		final int numAttributes = individuals.get(0).numAttributes();
		for (int attribute = 0; attribute < numAttributes; attribute++) {
			wrappedIndividuals.sort(new ScoreComparator(attribute));

			CrowdingDistanceIndividual min = wrappedIndividuals.get(0);
			CrowdingDistanceIndividual max = wrappedIndividuals.get(lastIndex);

			final double minValue = min.getIndividual().getScore(attribute);
			final double maxValue = max.getIndividual().getScore(attribute);

			// no divergence?
			if (minValue >= maxValue) continue;

			// calculate crowding distance
			for (int i = 1; i < lastIndex; i++) {
				CrowdingDistanceIndividual current = wrappedIndividuals.get(i);
				CrowdingDistanceIndividual leftNeighbor = wrappedIndividuals.get(i - 1);
				CrowdingDistanceIndividual rightNeighbor = wrappedIndividuals.get(i + 1);

				// calculate & accumulate normalized crowding distance
				final double leftValue = leftNeighbor.getIndividual().getScore(attribute);
				final double rightValue = rightNeighbor.getIndividual().getScore(attribute);
				final double distance = (rightValue - leftValue) / (maxValue - minValue);
				current.addCrowdingDistance(distance);
			}

			// update extremes
			min.addCrowdingDistance(2.0);
			max.addCrowdingDistance(2.0);
		}

		// sort by descending crowding distance
		wrappedIndividuals.sort(Collections.reverseOrder(new CrowdingDistanceComparator()));

		// strip crowding distance wrapper
		List<Individual> result = new ArrayList<Individual>();
		for (CrowdingDistanceIndividual crowdingDistanceIndividual : wrappedIndividuals) {
			result.add(crowdingDistanceIndividual.getIndividual());
		}

		return result;
	}

	private final DominationComparator comparator = new DominationComparator();
}

File: csharp/InfiniteZelda/Mooga/CrowdingDistanceRanking.cs
using System;
using System.Collections.Generic;

namespace Lumpn.Mooga
{
    public sealed class CrowdingDistanceRanking : IRanking
    {
        public IEnumerable<Individual> Rank(IList<Individual> individuals)
        {
            // trivial?
            if (individuals.Count < 2)
            {
                return new List<Individual>(individuals);
            }

            // split into non-dominated and dominated
            List<Individual> nonDominated = new List<Individual>();
            List<Individual> dominated = new List<Individual>();
            foreach (Individual individual in individuals)
            {
                // check domination
                bool isDominated = false;
                foreach (Individual other in individuals)
                {
                    if (other == individual) continue;
                    if (comparator.Compare(individual, other) < 0)
                    {
                        isDominated = true;
                        break;
                    }
                }

                // put in respective list
                if (isDominated)
                {
                    dominated.Add(individual);
                }
                else
                {
                    nonDominated.Add(individual);
                }
            }

            // sort non-dominated by crowding distance
            List<Individual> sortedNonDominated = sortByCrowdingDistance(nonDominated);

            // recursively rank the dominated individuals
            List<Individual> rankedDominated = Rank(dominated);

            // concatenate both lists
            List<Individual> result = new List<Individual>();
            result.AddRange(sortedNonDominated);
            result.AddRange(rankedDominated);
            return result;
        }

        private List<Individual> SortByCrowdingDistance(List<Individual> individuals)
        {
            // trivial?
            if (individuals.Count < 2)
            {
                return new List<Individual>(individuals);
            }

            // create crowding distance wrapper
            List<CrowdingDistanceIndividual> wrappedIndividuals = new List<CrowdingDistanceIndividual>();
            foreach (Individual individual in individuals)
            {
                wrappedIndividuals.Add(new CrowdingDistanceIndividual(individual));
            }

            // calculate each crowding distance
            int lastIndex = wrappedIndividuals.Count - 1;
            int numAttributes = individuals[0].NumAttributes;
            for (int attribute = 0; attribute < numAttributes; attribute++)
            {
                wrappedIndividuals.Sort(new ScoreComparator(attribute));

                CrowdingDistanceIndividual min = wrappedIndividuals[0];
                CrowdingDistanceIndividual max = wrappedIndividuals[lastIndex];

                double minValue = min.individual.Score(attribute);
                double maxValue = max.individual.Score(attribute);

                // no divergence?
                if (minValue >= maxValue) continue;

                // calculate crowding distance
                for (int i = 1; i < lastIndex; i++)
                {
                    CrowdingDistanceIndividual current = wrappedIndividuals[i];
                    CrowdingDistanceIndividual leftNeighbor = wrappedIndividuals[i - 1];
                    CrowdingDistanceIndividual rightNeighbor = wrappedIndividuals[i + 1];

                    // calculate & accumulate normalized crowding distance
                    double leftValue = leftNeighbor.individual.Score(attribute);
                    double rightValue = rightNeighbor.individual.Score(attribute);
                    double distance = (rightValue - leftValue) / (maxValue - minValue);
                    current.AddCrowdingDistance(distance);
                }

                // update extremes
                min.AddCrowdingDistance(2.0);
                max.AddCrowdingDistance(2.0);
            }

            // sort by descending crowding distance
            wrappedIndividuals.Sort(new CrowdingDistanceComparator());
            wrappedIndividuals.Reverse();

            // strip crowding distance wrapper
            List<Individual> result = new List<Individual>();
            foreach (CrowdingDistanceIndividual crowdingDistanceIndividual in wrappedIndividuals)
            {
                result.Add(crowdingDistanceIndividual.individual);
            }

            return result;
        }

        private readonly DominationComparator comparator = new DominationComparator();
    }
}